File: zig/dnd-character/dnd_character.zig
const std = @import("std");
const RanGen = std.rand.DefaultPrng;
pub fn modifier(score: i8) i8 {
    return @divFloor((score - 10), 2);
}

pub fn ability() i8 {
    const seed: u64 = @bitCast(std.time.timestamp());
    var rnd: std.rand.Xoshiro256 = RanGen.init(seed);
    const n1 = rnd.random().intRangeLessThan(i8, 1, 7);
    const n2 = rnd.random().intRangeLessThan(i8, 1, 7);
    const n3 = rnd.random().intRangeLessThan(i8, 1, 7);
    const n4 = rnd.random().intRangeLessThan(i8, 1, 7);

    const arr: []const i8 = &[4]i8{ n1, n2, n3, n4 };

    var min_num: i8 = 7;
    for (arr) |dice| {
        if (dice < min_num) {
            min_num = dice;
        }
    }

    return n1 + n2 + n3 + n4 - min_num;
}

pub const Character = struct {
    strength: i8,
    dexterity: i8,
    constitution: i8,
    intelligence: i8,
    wisdom: i8,
    charisma: i8,
    hitpoints: i8,

    pub fn init() Character {
        const strength = ability();
        const dexterity = ability();
        const constitution = ability();
        const intelligence = ability();
        const wisdom = ability();
        const charisma = ability();
        const hitpoints = 10 + modifier(constitution);

        return Character{ .strength = strength, .dexterity = dexterity, .constitution = constitution, .intelligence = intelligence, .wisdom = wisdom, .charisma = charisma, .hitpoints = hitpoints };
    }
};

File: java/dnd-character/src/main/java/DnDCharacter.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

class DnDCharacter {

  private int strength;
  private int dexterity;
  private int consitution;
  private int intelligence;
  private int wisdom;
  private int charisma;
  private int charModifier;
  private int hitpoints;

  public DnDCharacter() {

    this.strength = ability(rollDice());
    this.dexterity = ability(rollDice());
    this.consitution = ability(rollDice());
    this.intelligence = ability(rollDice());
    this.wisdom = ability(rollDice());
    this.charisma = ability(rollDice());
    this.charModifier = modifier(consitution);
    this.hitpoints = 10 + this.charModifier;
  }

  int ability(List<Integer> scores) {
    int min = 7;
    int sum = 0;

    for (int i = 0; i < scores.size(); i++) {
      int currentValue = scores.get(i);
      if (currentValue < min) {
        min = currentValue;
      }

      sum += currentValue;
    }

    return sum - min;
  }

  List<Integer> rollDice() {
    Random rand = new Random();

    List<Integer> dices = new ArrayList<>();

    for (int i = 0; i < 4; i++) {
      dices.add((int) (rand.nextInt(6) + 1));
    }

    return dices;
  }

  int modifier(int input) {
    return Math.floorDiv(input - 10, 2);
  }

  int getStrength() {
    return this.strength;
  }

  int getDexterity() {
    return this.dexterity;
  }

  int getConstitution() {
    return this.consitution;
  }

  int getIntelligence() {
    return this.intelligence;
  }

  int getWisdom() {
    return this.wisdom;
  }

  int getCharisma() {
    return this.charisma;
  }

  int getHitpoints() {
    return this.hitpoints;
  }
}
